game/src/edit/roads.rs

```diff
@@ -1,84 +1,461 @@
+                        make_main_panel(ctx, app, app.primary.map.get_r(self.r), self.current_lane);
+                    self.highlight_selection =
+                        highlight_current_selection(ctx, app, self.r, self.current_lane);
+                } else if x == "delete lane" {
+                    self.modify_current_lane(ctx, app, None, |new, idx| {
+                        new.lanes_ltr.remove(idx);
+                    });
+                } else if x == "flip direction" {
+                    self.modify_current_lane(ctx, app, Some(0), |new, idx| {
+                        new.lanes_ltr[idx].dir = new.lanes_ltr[idx].dir.opposite();
+                    });
+                } else if x == "move lane left" {
+                    self.modify_current_lane(ctx, app, Some(-1), |new, idx| {
+                        new.lanes_ltr.swap(idx, idx - 1);
+                    });
+                } else if x == "move lane right" {
+                    self.modify_current_lane(ctx, app, Some(1), |new, idx| {
+                        new.lanes_ltr.swap(idx, idx + 1);
+                    });
+                } else if let Some(lt) = x.strip_prefix("change to ") {
+                    let lt = LaneType::from_short_name(lt).unwrap();
+                    self.modify_current_lane(ctx, app, Some(0), |new, idx| {
+                        new.lanes_ltr[idx].lt = lt;
+                    });
+                } else if let Some(lt) = x.strip_prefix("add ") {
+                    let lt = LaneType::from_short_name(lt).unwrap();
+
+                    let mut edits = app.primary.map.get_edits().clone();
+                    edits
+                        .commands
+                        .push(app.primary.map.edit_road_cmd(self.r, |new| {
+                            new.lanes_ltr.push(LaneSpec {
+                                lt,
+                                dir: Direction::Fwd,
+                                width: NORMAL_LANE_THICKNESS,
+                            });
+                        }));
+                    apply_map_edits(ctx, app, edits);
+                    self.redo_stack.clear();
+
+                    assert!(self.current_lane.is_none());
+                    self.current_lane =
+                        Some(app.primary.map.get_r(self.r).lanes_ltr().last().unwrap().0);
+                    self.main_panel =
+                        make_main_panel(ctx, app, app.primary.map.get_r(self.r), self.current_lane);
+                    self.highlight_selection =
+                        highlight_current_selection(ctx, app, self.r, self.current_lane);
+                    self.top_panel = make_top_panel(
+                        ctx,
+                        app,
+                        self.num_edit_cmds_originally,
+                        self.redo_stack.is_empty(),
+                    );
+                } else {
+                    unreachable!()
+                }
+            }
+            Outcome::Changed => {
```

Comment:
It's always been a bit weird to me that we don't have info about what changed. 

I guess we don't necessarily need it, like we do with clicked, but it feels a bit strange to "recompute everything" which is changeable.

No change needed now, just airing my widgetry grievances. 😬

Reply:
I'll do this refactor next; we can just plumb the name of the dropdown/slider/whatever else changed, and the caller can call `dropdown_value` and have the generic type resolution work out